# EV Population Data - Audit

**Purpose:**
This notebook inspects the raw EV dataset to understand its structure, data quality issues, and its limitations. No data cleaning or analysis is performed here.

In [9]:
import pandas as pd

In [10]:
df = pd.read_csv("../data/raw/Electric_Vehicle_Population_Data.csv")
df.head()

VIN (1-10)     County      City State  Postal Code  Model Year   Make  \
0  5YJYGDEE8L   Thurston  Tumwater    WA      98501.0        2020  TESLA   
1  5YJXCAE2XJ  Snohomish   Bothell    WA      98021.0        2018  TESLA   
2  5YJ3E1EBXK       King      Kent    WA      98031.0        2019  TESLA   
3  7SAYGDEE4T       King  Issaquah    WA      98027.0        2026  TESLA   
4  WAUUPBFF9G       King   Seattle    WA      98103.0        2016   AUDI   

     Model                   Electric Vehicle Type  \
0  MODEL Y          Battery Electric Vehicle (BEV)   
1  MODEL X          Battery Electric Vehicle (BEV)   
2  MODEL 3          Battery Electric Vehicle (BEV)   
3  MODEL Y          Battery Electric Vehicle (BEV)   
4       A3  Plug-in Hybrid Electric Vehicle (PHEV)   

   Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0            Clean Alternative Fuel Vehicle Eligible           291.0   
1            Clean Alternative Fuel Vehicle Eligible           238.0   
2            Clean Alternative Fuel Vehicle Eligible           220.0   
3  Eligibility unknown as battery range has not b...             0.0   
4              Not eligible due to low battery range            16.0   

   Legislative District  DOL Vehicle ID             Vehicle Location  \
0                  35.0       124633715  POINT (-122.89165 47.03954)   
1                   1.0       474826075   POINT (-122.18384 47.8031)   
2                  47.0       280307233  POINT (-122.17743 47.41185)   
3                  41.0       280786565   POINT (-122.03439 47.5301)   
4                  43.0       198988891  POINT (-122.35436 47.67596)   

                                Electric Utility  2020 Census Tract  
0                         PUGET SOUND ENERGY INC       5.306701e+10  
1                         PUGET SOUND ENERGY INC       5.306105e+10  
2  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303303e+10  
3  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
4   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303300e+10

In [11]:
df.shape
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 270262 entries, 0 to 270261
Data columns (total 16 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         270262 non-null  object 
 1   County                                             270252 non-null  object 
 2   City                                               270252 non-null  object 
 3   State                                              270262 non-null  object 
 4   Postal Code                                        270252 non-null  float64
 5   Model Year                                         270262 non-null  int64  
 6   Make                                               270262 non-null  object 
 7   Model                                              270262 non-null  object 
 8   Electric Vehicle Type                              270262 non-null  object

In [12]:
df.describe(include="all")

VIN (1-10)  County     City   State    Postal Code     Model Year  \
count       270262  270252   270252  270262  270252.000000  270262.000000   
unique       16415     242      864      51            NaN            NaN   
top     7SAYGDEE7P    King  Seattle      WA            NaN            NaN   
freq          1171  133903    42125  269613            NaN            NaN   
mean           NaN     NaN      NaN     NaN   98176.713849    2021.964468   
std            NaN     NaN      NaN     NaN    2569.741818       3.053960   
min            NaN     NaN      NaN     NaN    1030.000000    1999.000000   
25%            NaN     NaN      NaN     NaN   98052.000000    2021.000000   
50%            NaN     NaN      NaN     NaN   98133.000000    2023.000000   
75%            NaN     NaN      NaN     NaN   98382.000000    2024.000000   
max            NaN     NaN      NaN     NaN   99577.000000    2026.000000   

          Make    Model           Electric Vehicle Type  \
count   270262   270262                          270262   
unique      47      183                               2   
top      TESLA  MODEL Y  Battery Electric Vehicle (BEV)   
freq    111049    57335                          215859   
mean       NaN      NaN                             NaN   
std        NaN      NaN                             NaN   
min        NaN      NaN                             NaN   
25%        NaN      NaN                             NaN   
50%        NaN      NaN                             NaN   
75%        NaN      NaN                             NaN   
max        NaN      NaN                             NaN   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
count                                              270262   270257.000000   
unique                                                  3             NaN   
top     Eligibility unknown as battery range has not b...             NaN   
freq                                               169872             NaN   
mean                                                  NaN       40.386332   
std                                                   NaN       79.342202   
min                                                   NaN        0.000000   
25%                                                   NaN        0.000000   
50%                                                   NaN        0.000000   
75%                                                   NaN       33.000000   
max                                                   NaN      337.000000   

        Legislative District  DOL Vehicle ID             Vehicle Location  \
count          269613.000000    2.702620e+05                       270174   
unique                   NaN             NaN                         1080   
top                      NaN             NaN  POINT (-122.13158 47.67858)   
freq                     NaN             NaN                         6588   
mean               28.850107    2.441199e+08                          NaN   
std                14.895435    6.430872e+07                          NaN   
min                 1.000000    4.385000e+03                          NaN   
25%                17.000000    2.194414e+08                          NaN   
50%                32.000000    2.615051e+08                          NaN   
75%                42.000000    2.776210e+08                          NaN   
max                49.000000    4.791150e+08                          NaN   

                                     Electric Utility  2020 Census Tract  
count                                          270252       2.702520e+05  
unique                                             77                NaN  
top     PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)                NaN  
freq                                            96367                NaN  
mean                                              NaN       5.297261e+10  
std                                               NaN       1.625614e+09  

In [13]:
df.isna().sum().sort_values(ascending=False)

Legislative District                                 649
Vehicle Location                                      88
County                                                10
City                                                  10
2020 Census Tract                                     10
Electric Utility                                      10
Postal Code                                           10
Electric Range                                         5
Model                                                  0
Make                                                   0
Model Year                                             0
State                                                  0
VIN (1-10)                                             0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Vehicle Type                                  0
DOL Vehicle ID                                         0
dtype: int64

In [14]:
df.duplicated().sum()

np.int64(0)

## Initial Observations

- Dataset contains ~270k EV registrations with 16 columns
- Each row represents a single registered electric vehicle
- Several location-related fields (Postal Code, Census Tract) are stored as floats, indicating missing values
- Electric Range has missing values, likely depending on vehicle type
- Model Year is available, but registration date is not, which limits true adoption trend analysis

## Missing Values & Duplicates

- Overall, missing data is minimal (<0.3% of rows)
- Legislative District has the highest missing count (649), likely due to administrative boundaries
- Location-related fields have small, consistent missing counts (10–88 rows)
- Electric Range is missing for a small number of vehicles
- No duplicate rows were detected, indicating strong data integrity
